compiler/rustc_hir_typeck/src/method/probe.rs

```diff
@@ -76,8 +77,12 @@ pub(crate) struct ProbeContext<'a, 'tcx> {
     /// requested name (by edit distance)
     allow_similar_names: bool,
 
+    /// List of potential private candidates. Will be trimmed to ones that
+    /// actually apply and then the result inserted into `private_candidate`
+    private_candidates: Vec<Candidate<'tcx>>,
+
     /// Some(candidate) if there is a private candidate
-    private_candidate: Option<(DefKind, DefId)>,
+    private_candidate: Cell<Option<(DefKind, DefId)>>,
```

Comment:
What is the difference between these lol

Reply:
Or actually, better question: Why do we need two fields here?

Reply:
We need the `private_candidate` field to communicate an actually found candidate out to the `pick` method. While we could just eagerly return the `PrivateMatch` error, that would mean we'd miss out on suggesting traits that could be imported to get access to another method of the same name.

Of course we could just eagerly report (not return) an error from method selection and return that no methods were found, which will subsequently return an error containing the list of traits that should be imported. But then method selection is reporting errors, which may also not be desirable.